<a href="https://colab.research.google.com/github/henryspg/DS-Unit-1-Sprint-2-Statistics/blob/master/Thursday_Statistics_Q%26A_DS16.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

# Set Pandas Display Options
pd.set_option('display.max_rows', 400)

In [0]:
# !curl https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data

In [0]:
df = pd.read_csv('auto-mpg.csv', na_values='?')

print(df.shape)
df.head()

(398, 9)


mpg  cylinders  displacement  ...  model year  origin                   car name
0  18.0          8         307.0  ...          70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...          70       1          buick skylark 320
2  18.0          8         318.0  ...          70       1         plymouth satellite
3  16.0          8         304.0  ...          70       1              amc rebel sst
4  17.0          8         302.0  ...          70       1                ford torino

[5 rows x 9 columns]

In [0]:
df.corr()

mpg  cylinders  ...  model year    origin
mpg           1.000000  -0.775396  ...    0.579267  0.563450
cylinders    -0.775396   1.000000  ...   -0.348746 -0.562543
displacement -0.804203   0.950721  ...   -0.370164 -0.609409
horsepower   -0.778427   0.842983  ...   -0.416361 -0.455171
weight       -0.831741   0.896017  ...   -0.306564 -0.581024
acceleration  0.420289  -0.505419  ...    0.288137  0.205873
model year    0.579267  -0.348746  ...    1.000000  0.180662
origin        0.563450  -0.562543  ...    0.180662  1.000000

[8 rows x 8 columns]

In [0]:
# Works if you end up dropping less tha 10% of your dataset
# df = df.dropna()

# print(df.shape)
# df.head()

In [0]:
# A pandas column is a numpy array

# df['mpg'] + df['cylinders']

In [0]:
df['horsepower'][0]

130.0

In [0]:
# for i, item in enumerate(df['horsepower']):
#   print(i)
#   int(item)

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 28.1+ KB


In [0]:
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model year        int64
origin            int64
car name         object
dtype: object

In [0]:
df.describe()

mpg   cylinders  ...  model year      origin
count  398.000000  398.000000  ...  398.000000  398.000000
mean    23.514573    5.454774  ...   76.010050    1.572864
std      7.815984    1.701004  ...    3.697627    0.802055
min      9.000000    3.000000  ...   70.000000    1.000000
25%     17.500000    4.000000  ...   73.000000    1.000000
50%     23.000000    4.000000  ...   76.000000    1.000000
75%     29.000000    8.000000  ...   79.000000    2.000000
max     46.600000    8.000000  ...   82.000000    3.000000

[8 rows x 8 columns]

In [0]:
df['horsepower'].value_counts(dropna=False)

In [0]:
df.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [0]:
df['cylinders'].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

# T-tests on horsepower

In [0]:
df['horsepower'].mean()

104.46938775510205

In [0]:
sample = df.sample(30)

sample.head()

mpg  cylinders  ...  origin                   car name
39   14.0          8  ...       1  pontiac catalina brougham
67   11.0          8  ...       1            mercury marquis
383  38.0          4  ...       3                honda civic
3    16.0          8  ...       1              amc rebel sst
43   13.0          8  ...       1   ford country squire (sw)

[5 rows x 9 columns]

In [0]:
sample.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model year        int64
origin            int64
car name         object
dtype: object

## 1-sample t-test

1) Null Hypothesis: 

My sample mean is equal to my null hypothesis value (my concept of the population): 104.4

2) Alternative Hypothesis:

My sample mean is not equal to the null hypothesis value.

3) Confidence Level: 95%

In [0]:
stats.ttest_1samp(sample['horsepower'], 104.4, nan_policy='omit')

Ttest_1sampResult(statistic=1.3005782419490362, pvalue=0.2040059185864372)

In [0]:
sample['horsepower'].mean()

114.62068965517241

Conclusion: Bawsed on a t-statistic of 1.3 and a p-value of .204, I **fail to reject** the null hypothesis that our sample's horsepower is equal to 104.4

In [0]:
def confidence_interval(data, confidence=0.95):
  """
  Calculate a confidence interval around a sample mean for given data.
  Using t-distribution and two-tailed test, default 95% confidence. 
  
  Arguments:
    data - iterable (list or numpy array) of sample observations
    confidence - level of confidence for the interval
  
  Returns:
    tuple of (mean, lower bound, upper bound)
  """
  data = np.array(data)
  mean = np.mean(data)
  n = len(data)
  # pass extra parameter (ddof=1) to calculate **sample** standard deviation
  s = data.std(ddof=1)
  stderr = s / np.sqrt(n)
  # Lookup the t-statistic that corresponds to 95% area under the curve
  # for my given degrees of freedom: 49
  t = stats.t.ppf((1 + confidence) / 2.0, n - 1)
  margin_of_error = t*stderr
  return (mean - margin_of_error, mean, mean + margin_of_error)

In [0]:
sample_cleaned = sample['horsepower'].dropna()

confidence_interval(sample_cleaned)

(98.5231316548531, 114.62068965517241, 130.71824765549172)

In [0]:
stats.ttest_1samp(sample_cleaned, 131)

Ttest_1sampResult(statistic=-2.084260003123457, pvalue=0.046378319052318415)

Based on a t-statistic of 2.08 and a p-value of .046, I reject the null hypothesis that our population value is equal to 131, suggest the alternative that there is a statistically significant difference between 104.4 and 131 given a sample size of 30. 

### p-value: The probability that the null hypothesis is true, given the sample that we have collected.

p-value >= critical value (.05 or 1-confidence_level) we will **fail to reject** the null hypothesis

p-value < critical value we will **reject** the null hypothesis.

The average number of horsepower in a car is... 90 - True, or not True?

## 2-sample t-test

In [0]:
df.head()

mpg  cylinders  displacement  ...  model year  origin                   car name
0  18.0          8         307.0  ...          70       1  chevrolet chevelle malibu
1  15.0          8         350.0  ...          70       1          buick skylark 320
2  18.0          8         318.0  ...          70       1         plymouth satellite
3  16.0          8         304.0  ...          70       1              amc rebel sst
4  17.0          8         302.0  ...          70       1                ford torino

[5 rows x 9 columns]

In [0]:
df['cylinders'].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

In [0]:
eight = df[df['cylinders']==8]
four = df[df['cylinders']==4]

### Do cars with 8 cylinders get the same gas mileage as cars that have 4 cylinders?

1) Null Hypothesis:

The two sample means are equal (8 cylinder cars get the same gas mileage as 4 cylinder cars)

2) Alternative:

The two sample means are not equal (different gas mileage)

Confidence Level: 95%

In [0]:
stats.ttest_ind(eight['mpg'], four['mpg'])

Ttest_indResult(statistic=-23.993604714696918, pvalue=3.3188076271687665e-72)

Conclusion: Based on a t-statistic of -24, and a p-value of 0. We reject the null hypothesis that 8 cylinder and 4 cylinder cars get the same gas mileage.



In [0]:
eight['mpg'].mean()

14.963106796116508

In [0]:
four['mpg'].mean()

## Do the two samples have equal variance?

In [0]:
four['mpg'].var()

32.605883077368844

In [0]:
eight['mpg'].var()

8.044507900247483

The assumption of equal variance that needs to be satisfied when we run t-tests, is likely violated by these two samples

In [0]:
# Welch's T-test (Robust Method)
# Fix the assumption
# Makes it harder to find a statisticaly signifcant result
# We call the ability for a test to find statistical significance "Statistical Power"
# We satisfy the assumption, but we lose a little bit of our ability to find a significant result
# Our p-values will be a little bit bigger
stats.ttest_ind(eight['mpg'], four['mpg'], equal_var=False)

Ttest_indResult(statistic=-29.36466000454742, pvalue=6.96499988801485e-91)

In [0]:
# Bias-Variance Tradeoff. 

## Chi^2 Test

What kind of variables do we need to run a chi^2 test?

CATEGORICAL!!!

In [0]:
df.describe(exclude='number')

car name
count          398
unique         305
top     ford pinto
freq             6

In [0]:
df['origin'].value_counts()

1    249
3     79
2     70
Name: origin, dtype: int64

In [0]:
df['cylinders'].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

1) Null Hypothesis:

The two variables are independent (there is no relationship or association between the two

2) Alternative Hypothesis:

They are depdendent (This could mean that they're correlated, but that the only way that two variables could be dependent) 

3) Confidence Level: 95%

In [0]:
observed = pd.crosstab(df['origin'], df['cylinders'])

observed

cylinders  3   4  5   6    8
origin                      
1          0  72  0  74  103
2          0  63  3   4    0
3          4  69  0   6    0

In [0]:
observed_margins = pd.crosstab(df['origin'], df['cylinders'], margins=True)

observed_margins

cylinders  3    4  5   6    8  All
origin                            
1          0   72  0  74  103  249
2          0   63  3   4    0   70
3          4   69  0   6    0   79
All        4  204  3  84  103  398

In [0]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed)

print(chi2)
print(p_value)
print(dof)
print(expected)

180.10177141976712
9.800693325588576e-35
8
[[  2.50251256 127.6281407    1.87688442  52.55276382  64.43969849]
 [  0.70351759  35.87939698   0.52763819  14.77386935  18.11557789]
 [  0.79396985  40.49246231   0.59547739  16.67336683  20.44472362]]


Based on a chi2 statistic of 180, and a p-value of 0, we reject the null hypothesis that origin and number of cylinders are independent. And suggest the alternative, that there is an association between the two. 

## Chi^2 test with binned variables

YOU DO NOT NEED TO KNOW THIS FOR THE SPRINT CHALLENGE

In [0]:
# pd.cut()

# pd.qcut()

In [0]:
df['horsepower_bins'] = pd.qcut(df['horsepower'], 5)

In [0]:
df['weight_bins'] = pd.qcut(df['weight'], 5)

In [0]:
df.head()

mpg  cylinders  ...       weight_bins  horsepower_bins
0  18.0          8  ...  (3085.2, 3806.0]   (100.0, 140.0]
1  15.0          8  ...  (3085.2, 3806.0]   (140.0, 230.0]
2  18.0          8  ...  (3085.2, 3806.0]   (140.0, 230.0]
3  16.0          8  ...  (3085.2, 3806.0]   (140.0, 230.0]
4  17.0          8  ...  (3085.2, 3806.0]   (100.0, 140.0]

[5 rows x 11 columns]

In [0]:
observed = pd.crosstab(df['horsepower_bins'], df['weight_bins'])
observed

weight_bins      (1612.999, 2155.0]  ...  (3806.0, 5140.0]
horsepower_bins                      ...                  
(45.999, 72.0]                   64  ...                 0
(72.0, 88.0]                     14  ...                 0
(88.0, 100.0]                     2  ...                 0
(100.0, 140.0]                    0  ...                18
(140.0, 230.0]                    0  ...                62

[5 rows x 5 columns]

1) Null: horsepower and weight bins are independent

2) Alternative: dependent

3) 95% CL

In [0]:
stats.chi2_contingency(observed)

(505.50788689784844,
 2.2853864743926863e-97,
 16,
 array([[16.73469388, 15.89795918, 16.10714286, 16.5255102 , 16.73469388],
        [17.34693878, 16.47959184, 16.69642857, 17.13010204, 17.34693878],
        [15.30612245, 14.54081633, 14.73214286, 15.11479592, 15.30612245],
        [14.89795918, 14.15306122, 14.33928571, 14.71173469, 14.89795918],
        [15.71428571, 14.92857143, 15.125     , 15.51785714, 15.71428571]]))

Conclusion: We reject the null hypothesis that horsepower and weight are independent. 

## Slicing Dataframes

In [0]:
df.iloc[0:3,0:3]

mpg  cylinders  displacement
0  18.0          8         307.0
1  15.0          8         350.0
2  18.0          8         318.0

In [0]:
df.iloc[0:1,0:3]

mpg  cylinders  displacement
0  18.0          8         307.0

In [0]:
df.iloc[1:2,0:3]

mpg  cylinders  displacement
1  15.0          8         350.0

In [0]:
df.iloc[0:1,0:3] + df.iloc[1:2,0:3] 

mpg  cylinders  displacement
0  NaN        NaN           NaN
1  NaN        NaN           NaN

In [0]:
df = pd.DataFrame({'mean': [2,67,43], 'std_dev': [4,8,5], 'issue': ['water_project', 'handicapped-infants', 'immigration']})

df.head()

mean  std_dev                issue
0     2        4        water_project
1    67        8  handicapped-infants
2    43        5          immigration

### Hierarchical Index:

Nested Column headers

In [0]:
df = pd.DataFrame(np.random.randn(3, 8), index=['A', 'B', 'C'], columns=index)

NameError: ignored

## Work with a dataframe without column headers

In [0]:
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/car/car.data",  header=None)

df.head()

0      1  2  3      4     5      6
0  vhigh  vhigh  2  2  small   low  unacc
1  vhigh  vhigh  2  2  small   med  unacc
2  vhigh  vhigh  2  2  small  high  unacc
3  vhigh  vhigh  2  2    med   low  unacc
4  vhigh  vhigh  2  2    med   med  unacc

In [0]:
# grab column 0 
df[0]

0       vhigh
1       vhigh
2       vhigh
3       vhigh
4       vhigh
        ...  
1723      low
1724      low
1725      low
1726      low
1727      low
Name: 0, Length: 1728, dtype: object

In [0]:
# grab column 4
df[4]

0       small
1       small
2       small
3         med
4         med
        ...  
1723      med
1724      med
1725      big
1726      big
1727      big
Name: 4, Length: 1728, dtype: object

In [0]:
# You can add column headers after reading in the CSV
df.columns = ['one', 'two', 'three', 'four', 'five', 'six', 'seven']

df.head()

one    two three four   five   six  seven
0  vhigh  vhigh     2    2  small   low  unacc
1  vhigh  vhigh     2    2  small   med  unacc
2  vhigh  vhigh     2    2  small  high  unacc
3  vhigh  vhigh     2    2    med   low  unacc
4  vhigh  vhigh     2    2    med   med  unacc

In [0]:
df.dtypes

one      object
two      object
three    object
four     object
five     object
six      object
seven    object
dtype: object

Empty DataFrame
Columns: []
Index: []